Improve write speed of microSD card via SPI and FatFS

FRASTM opened the pull request:
This PR is adding a new function `sdhc_spi_write_multi` to increase the write operation on SD card using a SPI bus. This function is sending the CMD25 (**Write multiple blocks**) instead of looping on CMD24  (Write single blocks) like in the `sdhc_spi_write` function.
 when the nb of blocks to write is significant (more than 2).
   
There is a sample application coming from  @EminYagmahan to measure transfer speed.
This PR fixes the "write" part of the https://github.com/zephyrproject-rtos/zephyr/issues/22906.
 
Test using the ci-shield (PR #26741) with the following command:
`$ west build -b nucleo_f411re samples/subsys/fs/spi_sd/ -DSHIELD="ci_shield"`

Note that the **multiple block read** operation is already in the `sdhc_spi_read()` function which sends the SDHC_READ_MULTIPLE_BLOCK command

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Measurements reported in the issue #22906 with this PR shows a slight improvement of the write transfer speed. 

Note that multi block transfer is for more than 2 blocks (for 1 or 2 blocks, the single mode is kept).

pabigot commented on samples/subsys/fs/spi_sd/src/main.c:
```diff
@@ -0,0 +1,307 @@
+ *-------------------------------------------------------------------------
+ */
+#define TEST_FILE_SIZE 1024
+#define KILO 1000
```
This results in an unusual transfer size, neither 10^6 nor 2^20.  It could be more consistent with normal testing if the multiplier were 1024, so the total transfer size is 1 MiBy.

FRASTM replied:
OK to set blocks of 1024  x 1024 bytes

pabigot reviewed (dismissed):
This contains #26319 which is blocked for requested changes to be provided.  The new material here is being reviewed, but must be rebased on master once that PR goes in.  The NAK is related to this dependency.

Some style comments, but I don't immediately see anything of functional concern.

It might be worth going back sometime and coming up with a shared sample that does file system performance the same way, rather than have each sample do it independently.

FRASTM commented on subsys/disk/disk_access_spi_sdhc.c:
```diff
@@ -50,7 +50,8 @@ struct sdhc_spi_data {
 
 DEVICE_DECLARE(sdhc_spi_0);
 
-static inline const struct spi_config *sdhc_get_spi_cfg(const struct sdhc_spi_data *data)
```
ok

FRASTM commented on subsys/disk/disk_access_spi_sdhc.c:
```diff
@@ -781,6 +784,87 @@ static int sdhc_spi_write(struct sdhc_spi_data *data,
+	}
+
+	err = 0;
+error:
```
using label "exit"

FRASTM commented:
update after comments from @pabigot

pabigot commented:
Blocked until contained material from #26319 is merged.  I'll remove my NAK after this is rebased on master with that content.

vaussard commented:
Really sorry for being slow with #26319, I will try to address comments ASAP.
But otherwise, this looks fairly orthogonal to #26319, so maybe you can even base your PR directly on master if I am too slow?

henrikbrixandersen commented:
@FRASTM Please rebase on master and fix the merge conflicts.

FRASTM commented:
rebase

henrikbrixandersen commented on samples/subsys/fs/spi_sd/README.rst:
```diff
@@ -0,0 +1,35 @@
+Building and Running
+********************
+
+This sample can be built for a variety of boards, including ``ci-shield``.
```
I don't believe we have in-tree support for the `ci_shield`? Perhaps reference another board/shield?

FRASTM replied:
this  is the purpose of PR #26741

henrikbrixandersen replied:
Yes. I understand that, but we should not reference a shield not yet in the tree.

FRASTM replied:
removing any ref to the board

henrikbrixandersen commented on subsys/disk/disk_access_spi_sdhc.c:
```diff
@@ -748,6 +748,87 @@ static int sdhc_spi_write(struct sdhc_spi_data *data,
+		return err;
+	}
+
+	sdhc_spi_set_cs(data, 0);
```
This way of handling SPI SDHC CS was removed in commit c64e0ce923cde54415d203e447a1dc62894ed2b0.

FRASTM replied:
replaced with 	`spi_release(data->spi, &data->cfg);`

henrikbrixandersen replied:
No, `spi_release(data->spi, &data->cfg);` should only be used where the `disk_access_spi_sdhc.c` used to de-assert the CS line. I believe this call can be removed all together.

henrikbrixandersen commented on subsys/disk/disk_access_spi_sdhc.c:
```diff
@@ -748,6 +748,87 @@ static int sdhc_spi_write(struct sdhc_spi_data *data,
+
+	err = 0;
+exit:
+	sdhc_spi_set_cs(data, 1);
```
```suggestion
	spi_release(data->spi, &data->cfg);
```

FRASTM replied:
replaced with `spi_release(data->spi, &data->cfg);`

henrikbrixandersen commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -17,5 +17,6 @@ supported:
   - usb device
   - nvs
   - counter
+  - lptim
```
Was this change intended to be included in this PR?

FRASTM replied:
mismatch, 
Removed

henrikbrixandersen commented on subsys/disk/disk_access_spi_sdhc.c:
```diff
@@ -748,6 +748,88 @@ static int sdhc_spi_write(struct sdhc_spi_data *data,
+	if (err != 0) {
+		return err;
+	}
+
+	spi_release(data->spi, &data->cfg);
+
```
I don't believe you need to release the SPI controller here?

henrikbrixandersen replied:
@FRASTM ^ ?

FRASTM replied:
line removed

henrikbrixandersen replied:
And reintroduced?

erwango commented:
@FRASTM , please rebase

FRASTM committed:
disk: sdhc: use the multiple block transfer for writing

The sdhc_spi_write_multi function is sending the CMD25
for multiple block write operation on the SD card
instead of looping the CMD24. This is improving
the transfer speed compared to the sdhc_spi_write function
when the nb of blocks to write is significant (more than 2).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase
